### ISSUE TITLE
Non-masked + masked merge bug fix

### ISSUE THREAD

**Issue by codekansas** [fchollet/keras#3065]
This was pointed out in [the PR thread](https://github.com/fchollet/keras/pull/2413) after it was merged, I'm not sure what the best way to go about making another change would be so I'll put it here.

Reference [this commit](https://github.com/codekansas/keras/commit/96c84158a569cc2a0e3fe256178a9d5415ee7164).

**`keras/engine/topology.py`**

``` python
             masks = [K.expand_dims(m, 0) for m in mask if m is not None]
              return K.all(K.concatenate(masks, axis=0), axis=0, keepdims=False)
          elif self.mode == 'concat':
 -            masks = [K.ones_like(inputs[i][:-1]) if m is None else m for i, m in zip(inputs, mask)]
 -            expanded_dims = [K.expand_dims(m) for m in masks]
 -            concatenated = K.concatenate(expanded_dims, axis=self.concat_axis)
 +            masks = [K.ones_like(inputs[i]) if m is None else K.expand_dims(m) for i, m in enumerate(mask)]
 +            concatenated = K.concatenate(masks, axis=self.concat_axis)
              return K.all(concatenated, axis=-1, keepdims=False)
          elif self.mode in ['cos', 'dot']:
              return None
```

**`tests/keras/layers/test_core.py`**

``` python
     # two different types of merging
      merged_sum = merge([masked_a, masked_b], mode='sum')
      merged_concat = merge([masked_a, masked_b], mode='concat', concat_axis=1)
 +    merged_concat_normal = merge([masked_a, input_b], mode='concat', concat_axis=1)

      # test sum
      model_sum = Model([input_a, input_b], [merged_sum])

                          ...

      model_concat.compile(loss='mse', optimizer='sgd')
      model_concat.fit([rand(2,3), rand(2,3)], [rand(2,6)], nb_epoch=1)

 +    # test concatenation with no mask
 +    model_concat_normal = Model([input_a, input_b], [merged_concat_normal])
 +    model_concat.compile(loss='mse', optimizer='sgd')
 +    model_concat.fit([rand(2,3), rand(2,3)], [rand(2,6)], nb_epoch=1)
 +

  def test_merge_mask_3d():
      from keras.layers import Input, merge, Embedding, SimpleRNN
```

**Comment by RaffEdwardBAH** [comment]
Maybe just make this another new pull request so @fchollet sees it and can merge it?

**Comment by RaffEdwardBAH** [comment]
@codekansas if you don't intend to create a new pull request for this fix, would you mind if I did?

**Comment by codekansas** [comment]
@RaffEdwardBAH yes, I think you should

**Comment by braingineer** [comment]
did this ever get put into a pull request?

**Comment by RaffEdwardBAH** [comment]
Not yet, been busy

**Comment by braingineer** [comment]
I'll hop on it in soonish (tonight).

**Comment by RaffEdwardBAH** [comment]
Awesome!

**Comment by pdasigi** [comment]
@braingineer is this done? If not, would you mind if I make a PR?

**Comment by braingineer** [comment]
@pdasigi not yet. ran into turbulence on project last night and didn't get to it.  

if you do,  check the commit link in first post (reposted below).  there is a comment I made on it concerning additional cases. 

https://github.com/codekansas/keras/commit/96c84158a569cc2a0e3fe256178a9d5415ee7164

**Comment by pdasigi** [comment]
Got it! Thanks!

**Comment by pdasigi** [comment]
Submitted a PR: https://github.com/fchollet/keras/pull/3218
Can someone please review it?
